README.rst
|License: GPL v3| |PyPI3| |PyPI1|

PIL Text Scaler
===============
Module for automatically fitting a string of text inside of a specified area.

Usage
_____
Loading a Font
--------------
To use a font with this module, it first needs to be loaded. The function ``loadTTF`` is used to load a True Type Font file. It takes 4 arguments: the name to be used for the font, the path of the font, an optional encoding, an option to load the fast version of the font, an optional minimum size, and an optional maximum size.

.. code:: python

    import PTS

    PTS.loadTTF('arial', 'arial.ttf')

Scaling Text
------------
To figure out if a string of text will fit in an area with the current settings, all you need to do is use the main function of the module: ``fitText``. The function takes 5 arguments: the text to be fit, the width of the area in pixels, the height of the area in pixels, an optional font name (defaults to "consolas"), an optional minimum size to use (use None to try all available sizes), and an option to use fast fonts, when available. The function returns a tuple of the wrapped text, the font that worked, and the size of said font, in that order.

.. code:: python

    import PTS

    text = "This is a string of text to be fit."

    PTS.loadTTF('arial', 'arial.ttf') # Load the font
    result = PTS.fitText(text, 100, 500, 'arial', 23, fast = True)

Getting the Minimum and Maximum Text Sizes
------------------------------------------
If you would like to get information on the sizes that a font is using, you can use the ``getSize`` function. It takes a font name and returns a dictionary with size information about that font.

.. code:: python

    import PTS

    information = PTS.getSize('arial')

Changing the Minimum and Maximum Text Sizes
-------------------------------------------
If you would like to change the minimum and maximum text sizes (as well as the difference between each size the module tries) you can use the ``setSize`` function. This function will change an existing font or will change the defaults for new fonts if no font has been specified. It takes 4 parameters: minimum size (inclusive), the maximum size (exclusive), an optional step parameter which is used to tell it how for to space each valid size from each other, and the name of a font to change. The default for these values are 15, 35, 2, and None, respectively.

.. code:: python

    import PTS

    minimum = 30
    maximum = 60
    step = 2

    PTS.setSizes(30, 60, 2) # Changes the default for new fonts.
    PTS.loadTTF('arial', 'arial.ttf') # Will load using the new sizes.

Fast Fonts
----------
Fast fonts are a way to process the text data much faster. The downsides are that they are memory intensive, taking longer to load, and are slightly less accurate. Fonts that have characters that overlap will be less accurate, sometimes by a large number of pixels.

.. |License: GPL v3| image:: https://img.shields.io/badge/License-GPLv3-blue.svg
   :target: LICENSE.txt

.. |PyPI3| image:: https://img.shields.io/badge/pypi-1.4.8-blue.svg
   :target: https://pypi.org/project/PTS/1.4.8/

.. |PyPI1| image:: https://img.shields.io/badge/python-3.6+-brightgreen.svg
   :target: https://www.python.org/downloads/release/python-367/


CHANGELOG.md
**v1.4.8**
* Added thread safety to the font loading. This allows you to add requests for loading a font from multiple threads. The dictionaries containing the font data are locked before modification or access and released immediately after to allow the operations to occur in parallel outside of the modification.

**v1.4.7**
* Attempted to fix an issue that would occur in text with leading and trailing spaces by having them automatically stripped from the text.

**v1.4.6**
* Updated README for PyPI upload (finally).
* Added `MANIFEST.in`.
* Fixed module name in `setup.py`.

**v1.4.5**
* Fixed `setup.py`.
* Added new entries to `.gitignore`.

**v1.4.4**
* Added some additional docstrings.
* Fixed the date.

**v1.4.3**
* A specific font was giving me a weird error with Pillow so I went and added some code that should fix it.

**v1.4.2**
* Fixed error in setup if Pillow is not installed.

**v1.4.1**
* Changed `getSize` and `setSizes` to not use case-sensitive font names.

**v1.4.0**
* Changed font size to be specific to a particular font. If no font size is specified, the defaults will be used.
* Changed `setSizes` to `setSize` and changed its function.
* Added new function `getSize` that gives you size details of a font.
* Fixed bug in 1.3.1 that caused `setSizes` to fail.

**v1.3.1**
* Added `MAX_SIZE` field to `PTS.core`. Avoid importing it directly to be able to detect changes to the font size.

**v1.3.0**
* Added new option `preferUnwrapped` to `fitText`. If this is enabled then the function will try to find a size of the font where it does not wrap the text mid-word. Will return the standard if it cannot find any sizes that do not do this.

**v1.2.3**
* Changed the way `FontError` works so that you can now access the font that caused the error. Access is done through `FontError.font`.

**v1.2.2**
* Made adjustment to the core to allow fonts to keep their case when returned from `listFonts`.

**v1.2.1**
* Added new function `listFonts` that returns a list of registered fonts.

**v1.2.0**
* Added new options to some commands for the capability to use fast fonts, a slightly less accurate calculation method that is much faster than the more accurate version.
* Added .gitignore

**v1.1.1**
* Fixed bug caused by `MIN_SIZE` accidentally being deleted.

**v1.1.0**
* Implemented threading to make longer strings faster.
* Fixed an error with `setSizes` so that it actually modified the global variables.

**v1.0.1**
* Fixed version number.
* Added new function to core: `attemptFit`.

**v1.0.0**
* Initial release.


PTS/size.py
import threading

FAST_LOCK = threading.Lock()
FAST_FONTS = {}

def createFastFont(fonts):
    """
    Creates a fast version of the specified Font object. While slightly less
    accurate, it is extremly fast.
    """
    for font in fonts:
        if font not in FAST_FONTS:
            try:
                with FAST_LOCK:
                    FAST_FONTS[font] = {chr(x): font.getsize_multiline(chr(x)) for x in range(256)}
            except OSError as e:
                if str(e) == 'unknown freetype error':
                    with FAST_LOCK:
                        FAST_FONTS[font] = {chr(x): font.getsize_multiline(chr(x)) for x in range(256)}
                else:
                    raise

def getSizeFast(text, font):
    """
    Faster version of the function of a Font object used for getting the size of
    a string of text. If a fast version of the Font object does not exist, it
    will automatically use the slow version of the function.
    """
    if font in FAST_FONTS:
        ffont = FAST_FONTS[font]
        lines = tuple(getSizeLine(line, font, ffont) for line in text.split('\n'))
        return (max(line[0] for line in lines), sum(line[1] for line in lines) + (4 * len(lines)))
    else:
        return getSizeSlow(text, font)

def getSizeLine(line, font, fastFont):
    return sumSizes(fastFont[char] if char in fastFont else font.getsize_multiline(char) for char in line)

def getSizeSlow(text, font):
    return font.getsize_multiline(text)

def sumSizes(sizes):
    """
    Takes in an iterable of sizes and returns the sums of each for a line.
    """
    sizes = tuple(sizes)
    return (sum(x[0] for x in sizes), max(x[1] for x in sizes))


PTS/__init__.py
#!/usr/bin/env python
# -*- coding: latin-1 -*-
# Date Format: YYYY-MM-DD

"""
PIL Text Scaler:
    Automatically scale text in a PIL image to fit in an area.

https://github.com/TheElementalOfDestruction/pts
"""

# --- LICENSE.txt -----------------------------------------------------------------
#
#    Copyright 2021 Destiny Peterson and Liam Moriarty
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program. If not, see <http://www.gnu.org/licenses/>.

__author__ = 'Destiny Peterson (The Elemental of Destruction)'
__date__ = '2021-10-25'
__version__ = '1.4.7'

from PTS.core import listFonts, loadTTF, fitText, getSize, setSize
from PTS.errors import FontError


PTS/core.py
import collections
import threading

import PIL.Image
import PIL.ImageDraw
import PIL.ImageFont

from PTS.errors import FontError
from PTS.size import FAST_FONTS, createFastFont, getSizeFast, getSizeSlow

REGISTERED = []
FONTS = {}
SIZES = (
    33,
    31,
    29,
    27,
    25,
    23,
    21,
    19,
    17,
    15,
)

MAX_SIZE = 33
MIN_SIZE = 15
STEP = 2

REF_IMG = PIL.Image.new('RGB', (1, 1), (0, 0, 0))
REF_DRAW = PIL.ImageDraw.ImageDraw(REF_IMG)


def attemptFit(text, words, width, height, font, size, fast):
    textsize = getSizeFast if fast else getSizeSlow
    currentAttempt = ''
    failed = False
    wrapped = False
    totalsize = textsize(text, font)
    if totalsize[0] <= width and totalsize[1] <= height: # This text will already fit into the area.
        return (text, font, size, wrapped)
    elif totalsize[1] > height: # This font size is already too big.
        return False
    # Create a list of words and whether or not we should split them.
    splitWords = [(word, font.getsize(word)[0] > width) for word in words]
    for word in splitWords:
        if textsize(currentAttempt + ' ' + word[0], font)[0] > width: # If the current word will overflow the line, we need to try a few things.
            if word[1]: # Can it be split?
                wrapped = True
                currentAttempt += '\n' if textsize(currentAttempt + ' ', font)[0] > width else ' '
                for character in word[0]:
                    currentAttempt += ('\n' + character) if textsize(currentAttempt + character, font)[0] > width else character
            else:
                currentAttempt += '\n' + word[0]
        else:
            currentAttempt += (' ' if (len(currentAttempt) > 0 and currentAttempt[-1] != '\n') else '') + word[0]
        if textsize(currentAttempt, font)[1] > height: # If the current attempt is two tall, we have failed.
            return False
    return (currentAttempt, font, size, wrapped)

def listFonts():
    return tuple(x for x in REGISTERED)

def loadTTF(name, path, encoding = '', fast = False, min = None, max = None, step = None):
    """
    Loads the font from the specified path and stores it with the specified
    name.
    """
    if name.lower() not in FONTS:
        max = max or MAX_SIZE
        min = min or MIN_SIZE
        step = step or STEP
        if max == MAX_SIZE and min == MIN_SIZE and step == STEP:
            sizes = SIZES
        else:
            sizes = tuple(reversed(range(min, max, int(abs(step)))))

        REGISTERED.append(name)
        name = name.lower()
        FONTS[name] = {size: PIL.ImageFont.truetype(path, size, encoding = encoding) for size in sizes}
        FONTS[name]['path'] = path
        FONTS[name]['encoding'] = encoding
        FONTS[name]['fast'] = fast
        FONTS[name]['sizes'] = sizes
        FONTS[name]['minSize'] = min
        FONTS[name]['maxSize'] = max
        FONTS[name]['step'] = step
        if fast:
            createFastFont(FONTS[name.lower()][size] for size in sizes)

def fitText(text, width, height, fontName = 'consolas', minSize = None, fast = False, preferUnwrapped = True):
    """
    Attempts to fit the text into the specified area. Will shrink the text size
    if the current size fails until it is less than minSize. Returns a tuple of
    the automatically wrapped text, the font that worked, and the size of the
    font. Returns None if the function failed. Will strip leading and trailing
    text.

    :param fast:            Tells it whether to use a faster (but slightly less
                            accurate) algorithm to determine the text size.
    :param preferUnwrapped: Tells the function to prefer a version that doesn't
                            wrap the text midword.
    """
    if fontName.lower() not in FONTS:
        raise FontError(fontName)

    fontName = fontName.lower()

    font = FONTS[fontName]

    if minSize is None:
        minSize = font['minSize']

    text = text.strip()

    words = text.split(' ')

    ret = collections.deque()
    def attemptFitRet(text, words, width, height, font, size, fast, ret):
        ret.append(attemptFit(text, words, width, height, font, size, fast))

    threads = tuple(threading.Thread(target = attemptFitRet, args = (text, words, width, height, font[size], size, fast, ret), daemon = True) for size in font['sizes'] if size >= minSize)
    for thread in threads:
        thread.start()

    for thread in threads:
        thread.join()

    largest = (None, None, 0, False)
    for result in ret:
        if result:
            if result[2] > largest[2]:
                largest = result

    if preferUnwrapped:
        # We need to get the largest unwrapped version.
        unwrappedLargest = (None, None, 0, False)
        for result in ret:
            if result and not result[3]:
                if result[2] > unwrappedLargest[2]:
                    unwrappedLargest = result

        largest = unwrappedLargest if unwrappedLargest[0] else largest

    return None if largest[0] is None else largest

def getSize(fontName):
    """
    Returns the minium size, the maximum size, and the step for the specified font.
    """
    if fontName.lower() not in FONTS:
        raise FontError(fontName)
    font = FONTS[fontName.lower()]
    return {'min': font['minSize'], 'max': font['maxSize'], 'step': font['step']}

def setSize(min, max, step, fontName = None):
    """
    Changes the size used for a specific font. If none are specified, changes
    the defaults and doesn't affect any existing fonts.
    :param min:      Minumum text size to use (inclusive).
    :param max:      Maximim text size to use (exclusive).
    :param step:     The step between sizes.
    :param fontName: The font to change.
    """
    sizes = tuple(reversed(range(min, max, int(abs(step)))))
    if fontName:
        for size in FONTS[fontName.lower()]['sizes']:
            if FONTS[fontName.lower()][size] in FAST_FONTS:
                del FAST_FONTS[FONTS[fontName.lower()][size]]

        path = FONTS[fontName.lower()]['path']
        encoding = FONTS[fontName.lower()]['encoding']
        del FONTS[fontName.lower()]
        loadTTF(fontName, path, encoding, min, max, step)

    else:
        global MAX_SIZE, MIN_SIZE, SIZES, STEP
        MAX_SIZE = max
        MIN_SIZE = min
        SIZES = sizes
        STEP = step
